### DISCUSSION TITLE
Is it possible to use a custom URL and api-key .......

### DISCUSSION THREAD

**Discussion by 321who** [ricklamers/gpt-code-ui#47] [Ideas]
I have found that on the website only the key on the openai based url is available. and I have some third party keys like say Poe and they are not available on this website. So is it possible to change the url to customize it ........
              Sorry my grammar may not be very good, it's because I'm using a translation program.......

**Comment by ricklamers** [comment]
https://github.com/ricklamers/gpt-code-ui/blob/main/.env.example

Check out these configurable ENV vars 👍🏻

**Reply by 321who** [reply]
<img width="564" alt="屏幕截图 2023-07-15 164344" src="https://github.com/ricklamers/gpt-code-ui/assets/129286057/fef4a8d3-12ac-4edd-addf-9aa70bcaf72b">

When I type
pip install gpt-code-ui
gptcode
and pip install "numpy>=1.24,<1.25" "dateparser>=1.1,<1.2" "pandas>=1.5,<1.6" "geopandas>=0.13,<0.14" "PyPDF2>=3.0,<3.1" "pdfminer>=20191125 ,<20191200" "pdfplumber>=0.9,<0.10" "matplotlib>=3.7,<3.8"
The file it creates does not have .env in it. do I have to create it directly in the root directory where the project is located, or into workspace?

**Reply by ricklamers** [reply]
>in the root directory where the project is located

Correct!

**Comment by 321who** [comment]
<img width="611" alt="屏幕截图 2023-07-15 171511" src="https://github.com/ricklamers/gpt-code-ui/assets/129286057/4ac684ad-854d-40cd-98d9-bc466884ed87">
hi，I created an .env file and changed the url and api to the format on .env.example. And made sure it was usable. The site is not reporting errors but is stuck, showing these in the terminal .....
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='api.openai.com', port=443): Max retries exceeded with url: /v1/chat/ completions (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x000001E685B78590>, 'Connection to api.openai.com')) timed out. (connect timeout=None)'))

And 

0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
[2023-07-15 17:02:17,460] ERROR [gpt_code_ui.webapp.main:1414] Exception on /generate [POST]
Is there something wrong?

**Reply by ricklamers** [reply]
Are you able to connect to the OpenAI using a `curl` command?

https://platform.openai.com/examples/default-explain-code?lang=curl

For example does running this succeed?

**Reply by 321who** [reply]
yep， I can connect to and browse this site. But if it's on gptcode, nope

**Reply by 321who** [reply]
<img width="1015" alt="屏幕截图 2023-07-16 100713" src="https://github.com/ricklamers/gpt-code-ui/assets/129286057/e02c70dd-4ee8-48b5-9595-ee134a36889d">

Oh, it seems to be working now? I honestly have no idea what's going on. But thanks, it's certainly a perfect project. And can I please join your discord? I want to learn and know more about computers, but now it's invitation seems to be expired or closed ................

**Reply by ricklamers** [reply]
https://discord.gg/ZmTQwpkYu6 isn’t working for you? It should not be expired as it’s set to not expire. What error are you getting?

**Reply by 321who** [reply]
Oh, and when I click on the link on the project it keeps coming up with a failing message.Maybe there's something wrong with my browser proxy. I just added it from the discord client. Thanks for the heads up.
